dask_expr/_expr.py

```diff
@@ -789,42 +789,66 @@ def __dask_keys__(self):
         return [(self._name, i) for i in range(self.npartitions)]
 
     def substitute(self, old, new) -> Expr:
-        """Substitute specific `Expr` instances within `self`
+        """Substitute a specific term within the expression
+
+        Note that replacing non-`Expr` terms may produce
+        unexpected results, and is not recommended.
+        Substituting boolean values is not allowed.
 
         Parameters
         ----------
-        substitutions:
-            mapping old terms to new terms. Note that using
-            non-`Expr` keys may produce unexpected results,
-            and substituting boolean values is not allowed.
+        old:
+            Old term to find and replace.
+        new:
+            New term to replace instances of `old` with.
 
         Examples
         --------
         >>> (df + 10).substitute(10, 20)
         df + 20
         """
 
-        old = old._name if isinstance(old, Expr) else old
-
-        if self._name == old:
-            return new
+        # Check if we are replacing a literal
+        if isinstance(old, Expr):
+            substitute_literal = False
+            if self._name == old._name:
+                return new
+        else:
+            substitute_literal = True
+            if isinstance(old, bool):
+                raise TypeError("Arguments to `substitute` cannot be bool.")
 
         new_exprs = []
         update = False
         for operand in self.operands:
-            if (
-                not isinstance(operand, bool)
-                and ishashable(operand)
-                and not isinstance(operand, Expr)
-                and operand == old
-            ):
-                new_exprs.append(new)
-                update = True
-            elif isinstance(operand, Expr):
+            if isinstance(operand, Expr):
                 val = operand.substitute(old, new)
                 if operand._name != val._name:
                     update = True
                 new_exprs.append(val)
+            elif (
+                isinstance(self, Fused)
+                and isinstance(operand, list)
+                and all(isinstance(op, Expr) for op in operand)
+            ):
+                # Special handling for `Fused`.
+                # We make no promise to dive through a
+                # list operand in general, but NEED to
+                # do so for the `Fused.exprs` operand.
```

Comment:
I'm having trouble finding a simple reproducer for this. However, this bug does seem to bite us for the specific implementation of q7 used in the pola-rs/tpch repo.

Reply:
I'll try to take a look as well (I have some practice with coming up with reproducer from these queries.... annoying task), but should be good to merge as is